### PR TITLE
Bug fix/fix signed int overflow

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix various value overflows in RandomGenerator code that were reported by UBSan.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)

- [x] I ensured this code runs with ASan / TSan or other static verification tools

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5700/

### CODE REVIEW

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5731/

File: lib/Random/RandomGenerator.cpp [hunk 1]
```diff
@@ -510,74 +518,36 @@ void RandomGenerator::ensureDeviceIsInitialized() {
...
+  if (left == INT64_MIN) {
+    diff = right + static_cast<uint64_t>(INT64_MAX);
   } else {
-    uint64_t high = static_cast<uint64_t>(right);
-    uint64_t low = static_cast<uint64_t>(left);
-    uint64_t d = high - low;
-    value = left + static_cast<int64_t>(interval(d));
-    TRI_ASSERT(value >= left && value <= right);
-    return value;
+    diff = right - left;
```
Comment (maierlars) [inline lib/Random/RandomGenerator.cpp]: for this to be correct you have to first cast `right` and `left` to `uint64_t`.

File: lib/Random/RandomGenerator.cpp [hunk 2]
```diff
@@ -100,7 +100,10 @@ int32_t RandomDevice::random(int32_t left, int32_t right) {
     return static_cast<int32_t>(random());
   }
 
-  uint32_t range = static_cast<uint32_t>(right - left + 1);
+  TRI_ASSERT(right > left);
+  int64_t r = static_cast<int64_t>(right) - static_cast<int64_t>(left) + 1;
```
Comment (graetzer) [inline lib/Random/RandomGenerator.cpp]: if the assert holds wouldn't  `static_cast<int64_t>(right - left)+1` be sufficient

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5751/

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5761/

**Comment by jsteemann** [comment]
Tests blue.
@maierlars : r u happy with it now?

**Pull request merged**